### DISCUSSION TITLE
Implications of vertex attributes?

### DISCUSSION THREAD

**Discussion by jombo23** [GPUOpen-LibrariesAndSDKs/RadeonProRenderBlenderAddon#617] [Show and tell]
Hello, I'm wondering what the implications behind getting a real vertex attribute setup are.

Since they're pretty important to the project I am working on, I wanted to see if I could implement them myself instead of bugging you guys to do it for me (thats the point of open source software right?) I originally thought it was a limitation of prorender as there were only like 14 different types of material lookup values. Then I went and skimmed through the prorender SDK, saw that there were things called primvars, read some more, and more, and more, and have now implemented a basic attribute system where the attribute name is its primvar key (obviously numbers only). I did this because 1, it seemed easy, and 2, I wasnt sure how both python/prorender generally handles, and anyone here wants scene globals to be (if its even required). 

It is my understanding that named attributes would have to be assigned a key, and even though that key can be reused for different purposes on different prorender shapes, it would be impossible to reverse look them up without storing the ID->textual name somewhere. For my own purposes, I wouldnt mind just making some global array somewhere and loading it with said pairs, however I believe that sounds like bad practice, and even if I did make a PR for it, theres a snowballs chance in heck it would be rolled in.

The current implementation of vertex colors is also broken as far as I could find. I couldnt make vertex_value0/1/2/3 return anything besides 0.

In my setup i have reimplemented vertex colors as primvar 99, and both made that fall under the "vertex color" option in RPRLookup, and also have added a primvar lookup into the node as well, seen here:
![41554cd8-ad23-4305-9f85-33e78b537c99b9b6560812](https://user-images.githubusercontent.com/4957057/233931284-0e4339fc-910a-4daf-aac0-d57fe528e660.jpg)

Vertex colors in use seen here:
![82f55f04-16a7-4029-9e3b-07a85296cf45375e13d29d](https://user-images.githubusercontent.com/4957057/233931931-1fb48eea-d4c8-425d-9057-41ef59fc33e6.jpg)

In order to make it seamless with eevee, I added a basic implementation of shadernodeattribute as well, so you can give it the ID and pull the values back.

Example case on this object trail I made using geometry nodes and attributes. The vertex position, curve point position, and curve point radius are stored in attributes so I can shade the trail based on the normal to the camera to give it a flat appearance and center the fade out from white to blue.

![0171c664-2eca-4e3c-ace7-d82813758750322dd4357a](https://user-images.githubusercontent.com/4957057/233932727-cb6cfcc4-eb72-4069-94bb-33fe5e6d051c.jpg)

![79de5525-f2d3-4e12-b5b1-8774f4668eb617d9ecb9a6](https://user-images.githubusercontent.com/4957057/233933284-f8a3f633-99e2-44f5-aa31-5a28c22265b9.jpg)

![427b15b7-a73f-44ac-925c-2be47061cc7fa958b2a045](https://user-images.githubusercontent.com/4957057/233933743-b6fee158-32ea-4b5b-aaa7-274bdb871326.jpg)

Small render of the sample curvy file I've been working with for this material/geonode setup, now working in prorender.
![output2](https://user-images.githubusercontent.com/4957057/233941443-3930b8a9-ed3a-41f9-a063-28ef5ad582ab.gif)

Anywho, thats all I've really got to show. My question is, if some random idiot like me can stumble through a language he doesnt know(python), and manage to make blender attributes work in a minimal, but usable way, what are the chances of some proper attribute support being added to RPR? It's gotta be much easier for you guys to work with it than I.

**Comment by bsavery** [comment]
Hey there.  I would make a PR with what you have so far, describe the issue you're not sure how to fix, and me or @VascoPi can help give some suggestion.  Sounds like you're part of the way there already.  

I'm assuming you saw the tutorial on the RPR SDK https://github.com/GPUOpen-LibrariesAndSDKs/RadeonProRenderSDK/blob/c61d7d3d56b0953d052ad561259d58e9c9a96f8f/tutorials/37_primvar/main.cpp#L187 on primvars?  This is of course C++ but should help a bit.

**Reply by jombo23** [reply]
Well I already have the primvars working perfectly (for my setup). what I'm talking about is strictly implementing a system to go from blender attribute names to a primvar id on a global scale. I'm not exactly sure how to create a name/Id table that the entire scene can access when exporting the mesh, and then how the material nodes can access in order to be able to read the correct primvar.

Like I said in my setup I'm just using the id directly instead of names, but the names themselves are textual.

I could make a pr, but currently even though it's only about 40 lines, it's messy and I would need to clean it up first.

My feelings were kinda that if I've proven that the concept was possible, it would be more reasonable for me to ask for a proper implemention, and someone would have an easier time implementing it properly than trying to rework my chicken scratch.(figure a case of a surcharge for "the customer worked on it first")

Of course, If you would still like me to post my current implementation that I only have handling floats and vectors on vertices by primvar id, I'd be more than happy to oblige, just let me know :)

Also on the note of the sdk, even though I write embedded c and not cpp, it seems to be written very nice, and the examples are really lovely, and I have brought over a couple more basic things from it into my own local setup. To me it is far more comprehensible than python haha.

**Reply by bsavery** [reply]
Yeah.  I see what you're saying.  Basically like you said you have to save a map of ID -> Name somewhere probably on the rpr_context object. 
However, just thinking about it.  Why wouldn't you just use the existing Blender "Color Attribute" node.  

Either way, yeah I think making the PR (we won't merge it till cleaned up) or pointing me at the branch online I can give some suggestions.

**Reply by jombo23** [reply]
The color attribute node only applies to the objects active color attribute (as far as I'm aware) with geometry nodes (and I believe even shader nodes now too) you can create attributes at runtime which store a handful of different types. all of course easily emulated and masked by primvars. I just wasn't sure about the id/name map :)

So be it! I'll clean up my code a little and post what I've done at some point in time. I've also been making my changes based on the RPRBLND-2251 branch(which has now been merged woohoo) locally, and not on GitHub, so I'll have to rebase and make a branch for it. Eventually (tm)

**Reply by jombo23** [reply]
I think I understand how to add this to the context item after looking it over. I am currently working on a few projects I need to get done (with prorender) in a reasonable amount of time, but I'll take a good stab at implementing it properly :)

**Reply by jombo23** [reply]
Crossing my fingers I did it somewhat decently. Managed to add actual named attribute support :)  #618

**Reply by VascoPi** [reply]
Looks good, let me take a look.

**Reply by VascoPi** [reply]
Here are some updates, [branch](https://github.com/VascoPi/RadeonProRenderBlenderAddon/tree/vertex_attributes_1).
Some modifications were made to the current implementation. Still in progress.
Now rpr_shape stores all uv maps as a primvars except the primary uv map. The UVMap node is improved. Probably the UI for the secondary UV map might be removed after primvar is onboard.
The Vertex color attribute stores all the vertex color layers as primvar. The support of ShaderNodeVertexColor.
![blender_8LWvqYt8qX](https://user-images.githubusercontent.com/42581166/235452496-62f32bb5-c65b-4ae9-b3a6-78b1e9f68fb0.png)

There is still some issue while adding to mesh new attribute. In progress.
![image](https://user-images.githubusercontent.com/42581166/235451096-a382c815-7828-4c57-b075-10f43935d812.png)
